File: examples/cpp/exampleScrewTheoryTrajectory/exampleScrewTheoryTrajectory.cpp
// -*- mode:C++; tab-width:4; c-basic-offset:4; indent-tabs-mode:nil -*-

/**
 * @ingroup kinematics-dynamics-examples
 * \defgroup screwTheoryTrajectoryExample screwTheoryTrajectoryExample
 *
 * <b>Legal</b>
 *
 * Copyright: (C) 2018 Universidad Carlos III de Madrid;
 *
 * CopyPolicy: Released under the terms of the LGPLv2.1 or later, see license/LGPL.TXT
 *
 * <b>Building</b>
\verbatim
cd examples/cpp/exampleScrewTheoryTrajectory/
mkdir build; cd build; cmake ..
make -j$(nproc)
\endverbatim
 * <b>Running example with teoSim</b>
 * First we must run a YARP name server if it is not running in our current namespace:
\verbatim
[on terminal 1] yarp server
\endverbatim
 * What mostly changes is the library command line invocation. We also change the server port name. The following is an example for the simulated robot's right arm.
\verbatim
[on terminal 2] teoSim
[on terminal 3] yarpdev --device BasicCartesianControl --name /teoSim/leftArm/CartesianControl --kinematics teo-fixedTrunk-leftArm-fetch.ini --robot remote_controlboard --local /BasicCartesianControl/teoSim/leftArm --remote /teoSim/leftArm --angleRepr axisAngle
[on terminal 4] ./exampleScrewTheoryTrajectory
\endverbatim
 */

#include <memory>
#include <vector>

#include <yarp/os/Bottle.h>
#include <yarp/os/LogStream.h>
#include <yarp/os/Network.h>
#include <yarp/os/Property.h>
#include <yarp/os/ResourceFinder.h>
#include <yarp/os/Time.h>

#include <yarp/dev/DeviceDriver.h>
#include <yarp/dev/IControlLimits.h>
#include <yarp/dev/IControlMode.h>
#include <yarp/dev/IEncoders.h>
#include <yarp/dev/IPositionDirect.h>
#include <yarp/dev/PolyDriver.h>

#include <kdl/frames.hpp>
#include <kdl/utilities/utility.h>
#include <kdl/path_line.hpp>
#include <kdl/rotational_interpolation_sa.hpp>
#include <kdl/trajectory_segment.hpp>
#include <kdl/velocityprofile_trap.hpp>
#include <kdl/chainfksolverpos_recursive.hpp>

#include <ConfigurationSelector.hpp>
#include <MatrixExponential.hpp>
#include <ProductOfExponentials.hpp>
#include <ScrewTheoryIkProblem.hpp>

#include "TrajectoryThread.hpp"

constexpr auto DEFAULT_REMOTE_PORT = "/ur16esim";
constexpr auto DEFAULT_TRAJ_DURATION = 10.0;
constexpr auto DEFAULT_TRAJ_MAX_VEL = 0.05;
constexpr auto DEFAULT_PERIOD_MS = 50.0;

namespace rl = roboticslab;

namespace
{
    KDL::Chain makeUR16eFromDh()
    {
        const KDL::Joint rotZ(KDL::Joint::RotZ);
        KDL::Chain chain;

        chain.addSegment(KDL::Segment(KDL::Joint(KDL::Joint::None), KDL::Frame(KDL::Rotation::RotZ(0))));

        //                                                     a       alpha      d    theta
        chain.addSegment(KDL::Segment(rotZ, KDL::Frame::DH(  0.0, -KDL::PI_2, 0.181,     0.0)));
        chain.addSegment(KDL::Segment(rotZ, KDL::Frame::DH(0.478,        0.0,   0.0,     0.0)));
        chain.addSegment(KDL::Segment(rotZ, KDL::Frame::DH( 0.36,        0.0,   0.0,     0.0)));
        chain.addSegment(KDL::Segment(rotZ, KDL::Frame::DH(  0.0, -KDL::PI_2, 0.174,     0.0)));
        chain.addSegment(KDL::Segment(rotZ, KDL::Frame::DH(  0.0,  KDL::PI_2,  0.12,     0.0)));
        // chain.addSegment(KDL::Segment(rotZ, KDL::Frame::DH(  0.0,        0.0,  0.19, KDL::PI)));
        chain.addSegment(KDL::Segment(rotZ, KDL::Frame::DH(  0.0,        0.0,  0.19,     0.0)));

        chain.addSegment(KDL::Segment(KDL::Joint(KDL::Joint::None), KDL::Frame(KDL::Rotation::RotX(KDL::PI_2))));

        return chain;
    }

    rl::PoeExpression makeUr16eKinematics()
    {
        KDL::Frame H_ST_0({0.838, 0.364, 0.061});
        rl::PoeExpression poe(H_ST_0);

        poe.append(rl::MatrixExponential(   rl::MatrixExponential::ROTATION, {0,  0, 1}, {    0,     0, 0.181}));
        poe.append(rl::MatrixExponential(   rl::MatrixExponential::ROTATION, {0,  1, 0}, {    0,     0, 0.181}));
        poe.append(rl::MatrixExponential(   rl::MatrixExponential::ROTATION, {0,  1, 0}, {0.478,     0, 0.181}));
        poe.append(rl::MatrixExponential(   rl::MatrixExponential::ROTATION, {0,  1, 0}, {0.838, 0.174, 0.181}));
        poe.append(rl::MatrixExponential(   rl::MatrixExponential::ROTATION, {0,  0,-1}, {0.838, 0.174, 0.061}));
        poe.append(rl::MatrixExponential(   rl::MatrixExponential::ROTATION, {0,  1, 0}, {0.838, 0.174, 0.061}));

        return poe;
    }
}

int main(int argc, char *argv[])
{
    yarp::os::Network yarp;

    if (!yarp::os::Network::checkNetwork())
    {
        yError() << "Please start a yarp name server first";
        return 1;
    }

    yarp::os::ResourceFinder rf;
    rf.configure(argc, argv);

    std::string remote = rf.check("remote", yarp::os::Value(DEFAULT_REMOTE_PORT), "remote port").asString();
    double trajDuration = rf.check("trajDuration", yarp::os::Value(DEFAULT_TRAJ_DURATION), "trajectory duration (s)").asFloat64();
    double trajMaxVel = rf.check("trajMaxVel", yarp::os::Value(DEFAULT_TRAJ_MAX_VEL), "trajectory max velocity (m/s)").asFloat64();
    int periodMs = rf.check("periodMs", yarp::os::Value(DEFAULT_PERIOD_MS), "command send period (ms)").asInt32();

    yarp::os::Property jointDeviceOptions {
        {"device", yarp::os::Value("remote_controlboard")},
        {"remote", yarp::os::Value(remote)},
        {"local", yarp::os::Value("/screwTheoryTrajectoryExample" + remote)}
    };

    yarp::dev::PolyDriver jointDevice(jointDeviceOptions);

    if (!jointDevice.isValid())
    {
        yError() << "Joint device not available";
        return 1;
    }

    yarp::dev::IEncoders * iEncoders;
    yarp::dev::IControlLimits * iControlLimits;
    yarp::dev::IControlMode * iControlMode;
    yarp::dev::IPositionDirect * iPositionDirect;

    if (!jointDevice.view(iEncoders) || !jointDevice.view(iControlLimits)
            || !jointDevice.view(iControlMode) || !jointDevice.view(iPositionDirect))
    {
        yError() << "Problems acquiring joint interfaces";
        return 1;
    }

    int axes;

    if (!iEncoders->getAxes(&axes))
    {
        yError() << "getAxes() failed";
        return 1;
    }

    std::vector<double> q(axes); // tested at: set poss (0 0 0 -90 0 0)

    while (!iEncoders->getEncoders(q.data()))
    {
        yarp::os::Time::delay(0.1);
    }

    // rl::PoeExpression poe = makeUr16eKinematics();

    // KDL::JntArray q0(6);
    // KDL::Frame H_ST_0;
    // poe.evaluate(q0, H_ST_0);

    // const auto chain = makeUR16eFromDh();
    // auto fkSolver = KDL::ChainFkSolverPos_recursive(chain);
    KDL::Frame H_DH_0;
    // fkSolver.JntToCart(q0, H_DH_0);

    // rl::PoeExpression poe2 = rl::PoeExpression::fromChain(makeUR16eFromDh());
    // poe2.evaluate(q0, H_DH_0);

    // yInfo() << "H_ST_0:" << H_ST_0.p.x() << H_ST_0.p.y() << H_ST_0.p.z();
    // yInfo() << "H_DH_0:" << H_DH_0.p.x() << H_DH_0.p.y() << H_DH_0.p.z();

    // yInfo() << "H_ST_0 == H_DH_0:" << (H_ST_0 == H_DH_0);

    // const auto chain = poe.toChain();

    // for (auto i = 0; i < chain.getNrOfSegments(); i++)
    // {
    //     const auto & segment = chain.getSegment(i);
    //     const auto H = segment.getFrameToTip();
    //     yInfo() << "Joint" << i << H.p.x() << H.p.y() << H.p.z();
    // }

    auto poe = rl::PoeExpression::fromChain(makeUR16eFromDh());

    // return 0;

    axes = poe.size(); // just for real TEO (7 joints, 6 motor axes)

    KDL::JntArray jntArray(axes);

    for (int i = 0; i < axes; i++)
    {
        jntArray(i) = KDL::deg2rad * q[i];
    }

    KDL::Frame H_base_start;

    if (!poe.evaluate(jntArray, H_base_start))
    {
        yError() << "FK error";
        return 1;
    }

    rl::ScrewTheoryIkProblemBuilder builder(poe);
    std::unique_ptr<rl::ScrewTheoryIkProblem> ikProblem(builder.build());

    if (!ikProblem.get())
    {
        yError() << "Unable to solve IK";
        return 1;
    }

    KDL::JntArray qMin(axes);
    KDL::JntArray qMax(axes);

    for (int i = 0; i < axes; i++)
    {
        if (!iControlLimits->getLimits(i, &qMin(i), &qMax(i)))
        {
            yError() << "Unable to retrieve limits for joint" << i;
            return 1;
        }
    }

    rl::ConfigurationSelectorLeastOverallAngularDisplacementFactory confFactory(qMin, qMax);
    std::unique_ptr<rl::ConfigurationSelector> ikConfig(confFactory.create());

    KDL::Frame H_base_end = H_base_start;
    H_base_end.p += {-0.3, 0.0, 0.0};

    KDL::RotationalInterpolation_SingleAxis interp;
    KDL::Path_Line path(H_base_start, H_base_end, &interp, 0.1, false);
    KDL::VelocityProfile_Trap profile(trajMaxVel, 0.2);
    KDL::Trajectory_Segment trajectory(&path, &profile, trajDuration, false);

    std::vector<int> modes(axes, VOCAB_CM_POSITION_DIRECT);

    if (!iControlMode->setControlModes(modes.data()))
    {
        yError() << "Unable to change mode";
        return 1;
    }

    TrajectoryThread trajThread(iEncoders, iPositionDirect, ikProblem.get(), ikConfig.get(), &trajectory, periodMs);

    if (trajThread.start())
    {
        yarp::os::Time::delay(trajDuration);
        trajThread.stop();
    }

    jointDevice.close();

    return 0;
}

File: examples/python/exampleScrewTheoryTrajectory.py
#!/usr/bin/env python3

import time
import numpy as np
import yarp
import PyKDL as kdl
import PySTS as sts
import threading
from motion_3 import *

# Placeholder imports for custom classes
# from your_module import PoeExpression, MatrixExponential, ScrewTheoryIkProblemBuilder, ConfigurationSelectorLeastOverallAngularDisplacementFactory, TrajectoryThread

DEFAULT_REMOTE_PORT = "/ur16esim"
DEFAULT_TRAJ_DURATION = 10.0
DEFAULT_TRAJ_MAX_VEL = 0.05
DEFAULT_PERIOD_MS = 50.0

def make_ur16e_kinematics():
    H_ST_0 = kdl.Frame(kdl.Vector(0.838, 0.364, 0.061))
    poe = sts.PoeExpression(H_ST_0)

    poe.append(sts.MatrixExponential(sts.MatrixExponential.motion.ROTATION, kdl.Vector(0, 0, 1), kdl.Vector(0, 0, 0.181)))
    poe.append(sts.MatrixExponential(sts.MatrixExponential.motion.ROTATION, kdl.Vector(0, 1, 0), kdl.Vector(0, 0, 0.181)))
    poe.append(sts.MatrixExponential(sts.MatrixExponential.motion.ROTATION, kdl.Vector(0, 1, 0), kdl.Vector(0.478, 0, 0.181)))
    poe.append(sts.MatrixExponential(sts.MatrixExponential.motion.ROTATION, kdl.Vector(0, 1, 0), kdl.Vector(0.838, 0.174, 0.181)))
    poe.append(sts.MatrixExponential(sts.MatrixExponential.motion.ROTATION, kdl.Vector(0, 0, -1), kdl.Vector(0.838, 0.174, 0.061)))
    poe.append(sts.MatrixExponential(sts.MatrixExponential.motion.ROTATION, kdl.Vector(0, 1, 0), kdl.Vector(0.838, 0.174, 0.061)))

    return poe

if __name__ == "__main__":
    yarp.Network.init()

    if not yarp.Network.checkNetwork():
        print("Please start a yarp name server first")
        exit(1)

    # Argument parsing (simplified)
    remote = DEFAULT_REMOTE_PORT
    traj_duration = DEFAULT_TRAJ_DURATION
    traj_max_vel = DEFAULT_TRAJ_MAX_VEL
    period_ms = DEFAULT_PERIOD_MS

    # Device options
    joint_device_options = yarp.Property()
    joint_device_options.put("device", "remote_controlboard")
    joint_device_options.put("remote", remote)
    joint_device_options.put("local", "/screwTheoryTrajectoryExample" + remote)

    joint_device = yarp.PolyDriver(joint_device_options)

    if not joint_device.isValid():
        print("Joint device not available")
        exit(1)

    iEncoders = joint_device.viewIEncoders()
    iControlLimits = joint_device.viewIControlLimits()
    iControlMode = joint_device.viewIControlMode()
    iPositionDirect = joint_device.viewIPositionDirect()

    if not (iEncoders and iControlLimits and iControlMode and iPositionDirect):
        print("Problems acquiring joint interfaces")
        exit(1)

    axes = iEncoders.getAxes()
    q = yarp.DVector(axes)

    while not iEncoders.getEncoders(q):
        time.sleep(0.1)

    poe = make_ur16e_kinematics()
    axes = poe.size()  # For real TEO (7 joints, 6 motor axes)

    jnt_array = kdl.JntArray(axes)
    for i in range(axes):
        jnt_array[i] = np.deg2rad(q[i])

    H_base_start = kdl.Frame()
    if not poe.evaluate(jnt_array, H_base_start):
        print("FK error")
        exit(1)

    builder = sts.ScrewTheoryIkProblemBuilder(poe)
    ik_problem = builder.build()

    if ik_problem is None:
        print("Unable to solve IK")
        exit(1)

    q_min = kdl.JntArray(axes)
    q_max = kdl.JntArray(axes)

    for i in range(axes):
        min, max = yarp.DVector(1), yarp.DVector(1)
        iControlLimits.getLimits(i, min, max)
        q_min[i] = math.radians(min[0])
        q_max[i] = math.radians(max[0])

    ik_config = sts.ConfigurationSelectorLeastOverallAngularDisplacement(q_min, q_max)

    H_base_end = kdl.Frame(H_base_start)
    H_base_end.p = H_base_end.p + kdl.Vector(-0.3, 0.0, 0.0)

    path = PathLine(H_base_start, H_base_end)
    profile = VelocityProfileRectangular(traj_max_vel)
    trajectory = TrajectorySegment(path, profile, traj_duration)

    if not iControlMode.setControlModes(yarp.IVector(axes, yarp.encode('posd'))):
        print("Unable to change mode")
        exit(1)

    should_stop = False

    def run():
        start_time = time.time()

        while not should_stop:
            movement_time = time.time() - start_time
            H_S_T = trajectory.position(movement_time)
            solutions, reachability = ik_problem.solve(H_S_T)

            # if not ik_config.configure(solutions, reachability):
            #     print("IK configuration failed")
            #     return

            # q = kdl.JntArray(axes)
            # q_deg = yarp.DVector(axes)

            # if not iEncoders.getEncoders(q_deg):
            #     print("getEncoders() failed")
            #     return

            # for i in range(q.rows()):
            #     q[i] = math.radians(q_deg[i])
            # print(q)
            # if not ik_config.findOptimalConfiguration(q):
            #     print("findOptimalConfiguration() failed")
            #     return

            # solution = ik_config.retrievePose()
            solution = solutions[0]

            refs = yarp.DVector([math.degrees(v) for v in solution])
            print(f"IK -> {list(refs)}")

            if not iPositionDirect.setPositions(refs):
                print("setPositions() failed")

    t = threading.Thread(target=run)
    t.start()

    time.sleep(traj_duration)
    should_stop = True
    t.join()

    joint_device.close()
    yarp.Network.fini()